### ISSUE TITLE
Major performance problems and hangs around if statements in BatchLexer

### ISSUE THREAD

**Issue by Anteru** [pygments/pygments-migration-test#923]
(Original issue 1216 created by chipx86 on 2016-03-01T10:04:03.766383+00:00)

We've hit an issue where batch files containing `if` statements can result in extremely long processing times, with the possibility of hanging indefinitely. It doesn't take very much to trigger this behavior. Simply create a batch file containing:

```
if "!abcdefghijklmnop!" equ "" echo hi
```

Or:

```
if "%abcdefghijklmnop%" equ "" echo hi
```

This will take so long that it might as well not finish at all. (It _does_ complete, but takes a while -- and I'll get into that more in a minute.)

There's no problem with this:

```
if "abcdefghijklmnop" equ "" echo hi
```

I spent a good amount of time digging into this. I can verify it was introduced in 2.1 with the `BatchLexer` work in 1bb2f7683bcb. The resulting regex that it's trying to process is:

```
((?:[&<>|]+|(?:(?:"[^\n\x1a"]*"?)|(?:(?:%(?:\*|(?:~[a-z]*(?:\$[^:]+:)?)?\d|[^%:\n\x1a]+(?::(?:~(?:-?\d+)?(?:,(?:-?\d+)?)?|(?:[^%\n\x1a^]|\^[^%\n\x1a])[^=\n\x1a]*=(?:[^%\n\x1a^]|\^[^%\n\x1a])*)?)?%))|(?:\^?![^!:\n\x1a]+(?::(?:~(?:-?\d+)?(?:,(?:-?\d+)?)?|(?:[^!\n\x1a^]|\^[^!\n\x1a])[^=\n\x1a]*=(?:[^!\n\x1a^]|\^[^!\n\x1a])*)?)?\^?!))|(?:(?:(?:\^[\n\x1a]?)?[^\n\x1a&<>|\t\v\f\r ,;=\xa0])+))+)(?:(?:(?:\^[\n\x1a])?[\t\v\f\r ,;=\xa0])+)?)(==)((?:(?:(?:\^[\n\x1a])?[\t\v\f\r ,;=\xa0])+)?(?:[&<>|]+|(?:(?:"[^\n\x1a"]*"?)|(?:(?:%(?:\*|(?:~[a-z]*(?:\$[^:]+:)?)?\d|[^%:\n\x1a]+(?::(?:~(?:-?\d+)?(?:,(?:-?\d+)?)?|(?:[^%\n\x1a^]|\^[^%\n\x1a])[^=\n\x1a]*=(?:[^%\n\x1a^]|\^[^%\n\x1a])*)?)?%))|(?:\^?![^!:\n\x1a]+(?::(?:~(?:-?\d+)?(?:,(?:-?\d+)?)?|(?:[^!\n\x1a^]|\^[^!\n\x1a])[^=\n\x1a]*=(?:[^!\n\x1a^]|\^[^!\n\x1a])*)?)?\^?!))|(?:(?:(?:\^[\n\x1a]?)?[^\n\x1a&<>|\t\v\f\r ,;=\xa0])+))+))
```

Okay, that's pretty unreadable, but it _looks_ like it's a `BatchLexer._stoken`.

Adding some further debugging, I see that it's trying to run the above sample code snippet through numerous regexes, with the above just being the most expensive of them. However, this is after it's already properly parsed all the previous bits of the code and generated syntax highlighting for them. I'm not sure how this particular parser works, so maybe that's normal, but it seems like it's probably not.

I've found that the longer the variable name in that string, the longer the run time. For instance, on my system, with variations of that string in a `.bat` file and running `time pygmentize test.bat`:

```
if "!abcdefghij!" equ "" echo hi
1.10s user 0.08s system 92% cpu 1.274 total

if "!abcdefghijk!" equ "" echo hi
1.98s user 0.06s system 98% cpu 2.078 total

if "!abcdefghijkl!" equ "" echo hi
3.68s user 0.07s system 99% cpu 3.774 total

if "!abcdefghijklm!" equ "" echo hi
7.02s user 0.10s system 98% cpu 7.196 total

if "!abcdefghijklmn!" equ "" echo hi
13.79s user 0.12s system 99% cpu 14.004 total

if "!abcdefghijklmno!" equ "" echo hi
27.44s user 0.22s system 99% cpu 27.873 total
```

Hope that info helps, and I'll be happy to provide any more information I can. I tried to come up with a fix, but I don't know the logic for this well enough yet.

**Comment by Anteru** [comment]
(Original issue 1216 last updated on 2016-03-02T07:39:02.028677+00:00)

(Issue automaticaly closed due to status in Bitbucket: resolved)

**Comment by Anteru** [comment]
(Original comment by dscorbett on 2016-03-01T13:14:09.009028+00:00)

I rewrote `BatchLexer` and introduced this problem. I’ll try to fix it later today.

**Comment by Anteru** [comment]
(Original comment by dscorbett on 2016-03-02T01:16:02.776056+00:00)

See pull request #563.

**Comment by Anteru** [comment]
(Original comment by birkenfeld on 2016-03-02T07:15:48.720055+00:00)

Merged, is released as version 2.1.3.

**Comment by Anteru** [comment]
(Original comment by chipx86 on 2016-03-02T07:39:02.022861+00:00)

Wow, thanks for taking care of all that so quickly! :)
